### PR TITLE
ARROW-16685: [Python] Preserve order of columns in joins

### PR DESCRIPTION
**Pull request by amol-**

### CODE REVIEW

**Review by jorisvandenbossche — APPROVED** [review]
Looks good!

File: python/pyarrow/_exec_plan.pyx [hunk 1]
```diff
@@ -259,13 +259,19 @@ def _perform_join(join_type, left_operand not None, left_keys,
         left_columns = []
     elif join_type == "inner":
         c_join_type = CJoinType_INNER
-        right_columns = set(right_columns) - set(right_keys)
+        right_columns = [
+            col for col in right_columns if col not in set(right_keys)
```
[the remark this answers is not in the document]
Comment (pitrou) [reply python/pyarrow/_exec_plan.pyx:263]:
```python
>>> %timeit [col for col in right_columns if col not in right_keys]
302 ns ± 2.81 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)
>>> %timeit [col for col in right_columns if col not in set(right_keys)]
630 ns ± 2.08 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)
>>> %timeit s = set(right_keys); [col for col in right_columns if col not in s]
370 ns ± 1.86 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)
```

[the remark this answers is not in the document]
Comment (amol-) [reply python/pyarrow/_exec_plan.pyx:263]: Given that it's not uncommon to see datasets with thousand of columns, especially in survey analytics, I went for using sets. But It's also true that the keys are probably going to be few even in that case.

Going for a list might introduce a slowdown in the case of very long column names, but I think we can live with that.

I'll switch to using a list

[the remark this answers is not in the document]
Comment (pitrou) [reply python/pyarrow/_exec_plan.pyx]: @amol- Hmm, please read what I wrote above.

[the remark this answers is not in the document]
Comment (amol-) [reply python/pyarrow/_exec_plan.pyx]: I was going to post some benchmarks in my previous answer, but I realised that in the end they didn't change much given that the keys will be in the order of single digit and the list was less code. So I went for the list. 

I don't have a strong opinion over this one, in all cases it looks like a minor change.

```
>>> import timeit
>>> right_columns =  ["a", "b", "c", "d", "e", "f"]
>>> right_keys = ["b", "c"]
>>> right_keys_s = set(right_keys)
>>> timeit.timeit(lambda: [col for col in right_columns if col not in set(right_keys)])
0.541847073007375
>>> timeit.timeit(lambda: [col for col in right_columns if col not in right_keys])
0.3103783940023277
>>> timeit.timeit(lambda: [col for col in right_columns if col not in right_keys_s])
0.236496517987689

>>> right_columns_long = ["panelist_question_answer_1", "panelist_question_answer_2", "panelist_question_answer_3", "panelist_question_answer_4", "panelist_question_answer_5", "panelist_question_answer_6"]
>>> right_keys_long = [ "panelist_question_answer_2", "panelist_question_answer_3" ]
>>> right_keys_long_s = set(right_keys_long)
>>> timeit.timeit(lambda: [col for col in right_columns_long if col not in set(right_keys_long)])
0.5470193719957024
>>> timeit.timeit(lambda: [col for col in right_columns_long if col not in right_keys_long])
0.32604770400212146
>>> timeit.timeit(lambda: [col for col in right_columns_long if col not in right_keys_long_s])
0.2378907290112693
```

[the remark this answers is not in the document]
Comment (amol-) [reply python/pyarrow/_exec_plan.pyx]: I think I found a good balance between the extra set creation and the using the existing list. I ended up reusing an existing dictionary keyed by join keys for the lookup. That is a tiny bit slower than using a set for the lookup but it's faster than list and doesn't involve any extra code.

Also if you account for the extra set creation it's actually faster
```
>>> timeit.timeit("right_keys_long_s = set(right_keys_long); [col for col in right_columns_long if col not in right_keys_long_s]", globals=globals())
0.30407846500747837
>>> timeit.timeit("[col for col in right_columns_long if col not in right_keys_long_d]", globals=globals())
0.24310287900152616
```

**Review by pitrou — APPROVED** [review]
Thanks for the update @amol-

**Pull request merged**